### PR TITLE
Add async support to batch operations

### PR DESCRIPTION
**Pull request by NetherGranite**
This adds async versions of all batch operations and their dependencies, and also adds async versions of all corresponding tests and their setup/teardown logic.

To avoid breaking changes, `IQueryProvider` is left untouched, and instead an `IAsyncQueryProvider` extension of it is added, and async batch operations look for an implementation of it in the provider collection, falling back to new async defaults if one isn't found. There is probably room for that fallback logic to be improved to be more configurable.

As a side effect, this bumps the C# version to 8. While C# 8 isn't fully supported by this project's target frameworks, C# 8 syntactic sugar is supported just fine. This was motivated by wanting `await using` to avoid very ugly code when multiple usings are used back-to-back.

Additionally, while .NET Framework supports `IAsyncDisposable`, it appears many classes only support it in .NET Standard, so a number of usings are wrapped in `#if`/`#else` blocks so that `await using` is only used if supported. See:

- [`DbCommand.DisposeAsync` Versions](https://learn.microsoft.com/en-us/dotnet/api/system.data.common.dbcommand.disposeasync#applies-to)
- [`DbDataReader.DisposeAsync` Versions](https://learn.microsoft.com/en-us/dotnet/api/system.data.common.dbdatareader.disposeasync#applies-to)
- [`IAsyncDisposable.DisposeAsync` Versions](https://learn.microsoft.com/en-us/dotnet/api/system.iasyncdisposable.disposeasync#applies-to)

### CODE REVIEW

**Comment by NetherGranite** [comment]
@RudeySH Aw man... I started this branch around a month ago and just got around to finishing it, but I see now that due to miraculous timing, there has been some refactoring done and effort has been made to add async support in the meantime. I would be happy to rebase this onto the refactored code on master if wanted for any reason, but I imagine your async work is much better than my naive copy/paste with changes to async calls.

**Comment by RudeySH** [comment]
Oh that's quite unfortunate, I've been working on the same thing. My changes haven't been published as a NuGet package yet, though, so the async methods aren't quite available yet. I'm intending to make breaking changes and so the new package will be a new major version, so I took a different approach and ended up just adding the async methods to `IQueryProvider`.

**Comment by NetherGranite** [comment]
@RudeySH It's awesome that you're still working on this project! It has solved some pretty critical issues for me 🙂

File: EntityFramework.Utilities/EntityFramework.Utilities/SqlQueryProvider.cs [hunk 1]
```diff
@@ -85,6 +86,47 @@ public void InsertItems<T>(IEnumerable<T> items, string schema, string tableName
 			}
 		}
 
+		public async Task InsertItemsAsync<T>(IEnumerable<T> items, string schema, string tableName, IList<ColumnMapping> properties, DbConnection storeConnection, int? batchSize, int? executeTimeout, SqlBulkCopyOptions copyOptions, DbTransaction transaction)
+		{
+			var reader = new EFDataReader<T>(items, properties);
+#if (NETSTANDARD2_1 || NETSTANDARD2_1_OR_GREATER || NETCOREAPP3_0_OR_GREATER)
```
Comment (RudeySH) [inline EntityFramework.Utilities/EntityFramework.Utilities/SqlQueryProvider.cs:92]: I'm copying some of these things over to my refactor. I was wondering about this right here. Wouldn't `#if (NETSTANDARD2_1_OR_GREATER)` be sufficient?

Comment (NetherGranite) [reply EntityFramework.Utilities/EntityFramework.Utilities/SqlQueryProvider.cs:92]: It turns out that the `*_OR_GREATER` is only supported on the .NET 5 SDK and higher, and given that this library is a little old, I wasn't sure what SDKs people might be using to develop this library, so I threw the extra flag in just to be safe. In modern Microsoft libraries, they definitely don't bother worrying about old SDKs. Totally up to you if you include it 🙂

**Pull request closed without merging**